### PR TITLE
feat: add metrics

### PR DESCRIPTION
**Pull request by randyahx**
### Description

Adding a metrics service  

### Rationale

To help monitor performance and detect errors  

### Example

none  

### Changes

Notable changes: 
none

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -3,6 +3,7 @@ package app
 import (
 	"encoding/json"
 	"fmt"
+	"github.com/bnb-chain/greenfield-challenger/metrics"
```
Comment (forcodedancing) [inline app/app.go:6]: Can you format the import?

**Comment by forcodedancing** [comment]
the lint ci fails

File: metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+const (
+	// Monitor
+	MetricGnfdSavedBlock = "gnfd_saved_block"
+	MetricGnfdSavedEvent = "gnfd_saved_event"
+	// increase(gnfd_saved_block[1m])
+	// increase(gnfd_saved_event[1m]) fix this
+
+	// Verifier
+	MetricVerifiedChallenges   = "verified_challenges"
+	MetricChallengeFailed      = "challenge_failed"
```
Comment (forcodedancing) [inline metrics/metrics.go:23]: Can we change the name `MetricChallengeFailed/Success` to make it more clear that the metrics is about `Verifier`?

Comment (forcodedancing) [inline metrics/metrics.go:23]: Maybe we can also have a metrics to record the latest block `Height`? Then we can compare it with fullnode, to see if there is any lag.

File: metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+		Name: MetricHashVerifierErr,
+		Help: "Hash verifier error count",
+	})
+	ms[MetricHashVerifierErr] = hashVerifierErrCountMetric
+	prometheus.MustRegister(hashVerifierErrCountMetric)
+
+	// Broadcaster
+	broadcasterErrCountMetric := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: MetricBroadcasterErr,
+		Help: "Succeeded challenges in database",
```
Comment (forcodedancing) [inline metrics/metrics.go:118]: Wrong description

File: metrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+	})
+	ms[MetricBroadcasterErr] = broadcasterErrCountMetric
+	prometheus.MustRegister(broadcasterErrCountMetric)
+
+	broadcastedChallengesMetric := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: MetricBroadcastedChallenges,
+		Help: "Broadcasted challengeID",
```
Comment (forcodedancing) [inline metrics/metrics.go:125]: Count of challenge?

File: metrics/metrics.go [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+	collatorErrCountMetric := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: MetricCollatorErr,
+		Help: "Collator error count",
+	})
+	ms[MetricCollatorErr] = collatorErrCountMetric
+	prometheus.MustRegister(collatorErrCountMetric)
+
+	collatedChallengesMetric := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: MetricCollatedChallenges,
+		Help: "Collated challengeID",
```
Comment (forcodedancing) [inline metrics/metrics.go:147]: Count of challenge?

File: metrics/metrics.go [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+	submitterErrCountMetric := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: MetricSubmitterErr,
+		Help: "Submitter error count",
+	})
+	ms[MetricSubmitterErr] = submitterErrCountMetric
+	prometheus.MustRegister(submitterErrCountMetric)
+
+	submitterChallengesMetric := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: MetricSubmittedChallenges,
+		Help: "Submitted challengeID",
```
Comment (forcodedancing) [inline metrics/metrics.go:169]: Count of challenge?

File: monitor/monitor.go [hunk 7]
```diff
@@ -1,6 +1,7 @@
 package monitor
 
 import (
+	"github.com/bnb-chain/greenfield-challenger/metrics"
```
Comment (forcodedancing) [inline monitor/monitor.go:4]: make format

File: metrics/metrics.go (hunk 2 above)
Comment (forcodedancing) [reply metrics/metrics.go:23]: It is there.

File: monitor/monitor.go [hunk 8]
```diff
@@ -154,7 +157,9 @@ func (m *Monitor) monitorChallengeEvents(block *tmtypes.Block, blockResults *cty
 	err = m.dataProvider.SaveBlockAndEvents(b, events)
 	for _, event := range events {
 		logging.Logger.Debugf("monitor event saved for challengeId: %d %s", event.ChallengeId, time.Now().Format("15:04:05.000000"))
+		m.metricService.SetGnfdSavedBlock(event.ChallengeId)
```
Comment (forcodedancing) [inline monitor/monitor.go:160]: SavedEvent, not SavedBlock here.

File: verifier/hash_verifier.go [hunk 9]
```diff
@@ -167,6 +171,7 @@ func (v *Verifier) verifyForSingleEvent(event *model.Event) error {
 	if challengeResErr != nil {
 		err = v.dataProvider.UpdateEventStatusVerifyResult(event.ChallengeId, model.Verified, model.HashMismatched)
 		if err != nil {
+			v.metricService.IncHashVerifierErr()
```
Comment (forcodedancing) [inline verifier/hash_verifier.go:174]: Can we have another metrics to record upstream challenge api issues - i.e., the count of errors when calling the api ? Then when clearly know whether the challenge api has issue or not.

File: metrics/metrics.go [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+	MetricBroadcasterErr        = "broadcaster_error_count"
+	// increase(broadcasted_challenges[3h]) check increase for heartbeat interval
+
+	// Vote Collector
+	// TODO: Do we need to check how many votes collected per challengeId?
+	// TODO: Should error be a counter or challengeId?
+	// Compare duration monitor/verifier verifier/broadcaster broadcaster/collator collator/submitter
+
+	// Vote Collator
+	MetricCollatedChallenges = "collated_challenges"
```
Comment (forcodedancing) [inline metrics/metrics.go:43]: Can we also add some metrics for vote collection?

File: vote/vote_collector.go [hunk 11]
```diff
@@ -14,18 +15,20 @@ import (
...
-	dataProvider DataProvider
+	config        *config.Config
+	executor      *executor.Executor
+	mtx           sync.RWMutex
+	dataProvider  DataProvider
+	metricService *metrics.MetricService
 }
 
-func NewVoteCollector(cfg *config.Config, executor *executor.Executor, collectorDataProvider DataProvider) *VoteCollector {
+func NewVoteCollector(cfg *config.Config, executor *executor.Executor, collectorDataProvider DataProvider, metricService *metrics.MetricService) *VoteCollector {
```
Comment (forcodedancing) [inline vote/vote_collector.go:25]: Can we also add some metrics for vote collection?

File: metrics/metrics.go (hunk 2 above)
Comment (randyahx) [reply metrics/metrics.go:23]: changed

File: metrics/metrics.go (hunk 3 above)
Comment (randyahx) [reply metrics/metrics.go:118]: changed

File: metrics/metrics.go (hunk 4 above)
Comment (randyahx) [reply metrics/metrics.go:125]: changed

File: metrics/metrics.go (hunk 5 above)
Comment (randyahx) [reply metrics/metrics.go:147]: changed

File: metrics/metrics.go (hunk 6 above)
Comment (randyahx) [reply metrics/metrics.go:169]: changed

File: monitor/monitor.go (hunk 8 above)
Comment (randyahx) [reply monitor/monitor.go:160]: changed

File: verifier/hash_verifier.go (hunk 9 above)
Comment (randyahx) [reply verifier/hash_verifier.go:174]: added

File: monitor/monitor.go (hunk 7 above)
Comment (randyahx) [reply monitor/monitor.go:4]: changed

File: app/app.go (hunk 1 above)
Comment (randyahx) [reply app/app.go:6]: changed

File: metrics/metrics.go (hunk 10 above)
Comment (randyahx) [reply metrics/metrics.go:36]: added

File: vote/vote_collector.go (hunk 11 above)
Comment (randyahx) [reply vote/vote_collector.go:25]: added

**Review by forcodedancing — APPROVED** [review]

**Pull request merged**